### PR TITLE
refactor: CI

### PR DESCRIPTION
**Pull request by gavindsouza**
### Changes

* Merge GitHub Actions manifests into one for Lint & Unit Tests jobs
* Update supported env list to [3.8, 3.9, 3.10, 3.11]
* Add Pre-commit File - Run `pre-commit install` to setup your hooks

Fixes https://github.com/kayak/pypika/issues/748, https://github.com/kayak/pypika/issues/747

### CODE REVIEW

File: .github/workflows/ci.yml [hunk 1]
```diff
@@ -1,19 +1,26 @@
...
 
 jobs:
-  test:
+  lint:
+    name: Lint
+    runs-on: ubuntu-latest
+    steps:
+      - uses: actions/checkout@v2
+      - uses: actions/setup-python@v2
+      - uses: psf/black@21.5b1
```
Comment (AzisK) [inline .github/workflows/ci.yml:15]: Can we have the same `black` version here as in `pre-commit-config.yaml`?

Comment (gavindsouza) [reply .github/workflows/ci.yml:15]: Forgot to replace this with the pre-commit invocation! Thanks for pointing it out.

**Comment by AzisK** [comment]
Good job Gavin. Could you also update the checkout Github Action `actions/checkout@v1` since there is already v4 out there?

**Comment by gavindsouza** [comment]
Hey @AzisK, is there anything else I can do to help get this merged?

**Comment by AzisK** [comment]
@gavindsouza it looks good but before that could you look at my last comments about `pypy` and `black` target version and `tox` envlist?

**Comment by gavindsouza** [comment]
> @gavindsouza it looks good but before that could you look at my last comments about `pypy` and `black` target version and `tox` envlist?

I can't find anything regarding those yet to be addressed. _(Did you add comments that are in a Pending state? If so, you are yet to submit your review)_

---

Meanwhile, since black is moved to pre-commit, I think removing it from dev-requirements tracking is best.

**Comment by AzisK** [comment]
@gavindsouza 

RE: _Meanwhile, since black is moved to pre-commit, I think removing it from dev-requirements tracking is best._

I agree 

RE: _I can't find anything regarding those yet to be addressed. (Did you add comments that are in a Pending state? If so, you are yet to submit your review)_

Yes, they have been in a pending state, sorry about that and thanks for the notice

File: .github/workflows/ci.yml [hunk 2]
```diff
@@ -1,32 +1,36 @@
-name: Unit Tests
+name: CI
 
 on:
   push:
-    branches: ["master"]
   pull_request:
-    branches: ["master"]
+  workflow_dispatch:
```
[the remark this answers is not in the document]
Comment (gavindsouza) [reply .github/workflows/ci.yml:6]: We can trigger tests on demand from the Actions tab with this - https://docs.github.com/en/actions/using-workflows/manually-running-a-workflow

File: tox.ini [hunk 3]
```diff
@@ -1,13 +1,13 @@
 [tox]
-envlist = py36,py37,py38,py39,pypy3
+envlist = py38,py39,py310,py311,py312,pypy3
```
[the remark this answers is not in the document]
Comment (gavindsouza) [reply tox.ini]: I've removed `3.12`. About the `pypy3` in the envlist, I'm not sure - I guess someone was using it in their development flows however I don't have a requirement on pypy so I haven't explored it

File: pyproject.toml [hunk 4]
```diff
@@ -1,8 +1,8 @@
 [tool.black]
 line-length = 120
 skip-string-normalization = true
-target-version = ['py36', 'py37', 'py38']
-exclude = '''
+target-version = ['py38', 'py39', 'py310', 'py311', 'py312']
```
[the remark this answers is not in the document]
Comment (gavindsouza) [reply pyproject.toml:4]:
```suggestion
target-version = ['py38', 'py39', 'py310', 'py311']
```

**Comment by AzisK** [comment]
Everything looks good, thanks! I will merge the pull request but version bump and release will happen a little bit later. However, this mostly concerns the CI, so I believe the version is not even very important in this case

**Pull request merged**